[inductor] Decompose torch.ops.quantized.embedding_bag_byte_unpack

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109398

This would be cleaner if we had support for u8->float32 views
(bitcasts) in inductor, but it works for now.

Differential Revision: [D49329910](https://our.internmc.facebook.com/intern/diff/D49329910/)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

bertmaher committed:
[inductor] Decompose torch.ops.quantized.embedding_bag_byte_unpack

This would be cleaner if we had support for u8->float32 views
(bitcasts) in inductor, but it works for now.

Differential Revision: [D49329910](https://our.internmc.facebook.com/intern/diff/D49329910/)

[ghstack-poisoned]

jgong5 commented on torch/_inductor/decomposition.py:
```diff
@@ -453,6 +453,17 @@ def dequantize_per_tensor_tensor_decomp_impl(
+        x, y, z, w = [u8[..., n].to(torch.int32) for n in (0, 1, 2, 3)]
+        return (x + (y << 8) + (z << 16) + (w << 24)).view(torch.float32)[..., None]
+
+    scales = bitcast_u8_to_f32(packed[..., -8:-4])
```
Directly view with `torch.float32` doesn't work? `packed[..., -8:-4].view(torch.float32)`.

bertmaher replied:
Unfortunately inductor doesn't support views with differing bitwidth yet, even though pytorch does.  You get a "bitcast to different bitwidth type is not supported yet" from lowering.py.  It would certainly be nice to support that at some point!

bertmaher committed:
support dynamic shapes on "[inductor] Decompose torch.ops.quantized.embedding_bag_byte_unpack"


This would be cleaner if we had support for u8->float32 views
(bitcasts) in inductor, but it works for now.

Differential Revision: [D49329910](https://our.internmc.facebook.com/intern/diff/D49329910/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bertmaher committed:
appease ruff on "[inductor] Decompose torch.ops.quantized.embedding_bag_byte_unpack"


This would be cleaner if we had support for u8->float32 views
(bitcasts) in inductor, but it works for now.

Differential Revision: [D49329910](https://our.internmc.facebook.com/intern/diff/D49329910/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bertmaher commented:
I suspect the breakages are present on trunk (or whatever base rev ghstack chose for this PR) 🤷
